### PR TITLE
Add database abstraction for history recorder

### PR DESCRIPTION
**Pull request by stealthycoin**
Adds database connection and interaction abstractions as well as a
handler that can hook into botocore's HistoryRecorder
 * DatabaseConnection
 * DatabaseRecordReader
 * DatabaseRecordWriter
 * PayloadSerializer
 * DatabaseHistoryHandler

cc @jamesls @kyleknap @dstufft @JordonPhillips @joguSD

### CODE REVIEW

File: tests/functional/history/test_db.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            {
+                'source': 'TEST',
+                'event_type': 'zip',
+                'payload': '',
+                'timestamp': 6
+            }
+        ])
+        reader = DatabaseRecordReader(self.connection)
+        cursor = self._get_cursor()
+        cursor.execute(
```
Comment (kyleknap) [inline tests/functional/history/test_db.py]: I wonder if it makes sense to expose the ability to supply your own id to the writer intializer instead of having to do this? Then you can test for the other id pretty easily as well.

File: tests/unit/customizations/history/test_db.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+class TestDatabaseRecordReader(BaseDatabaseRecordTester):
+    def setUp(self):
+        self.fake_connection = FakeDatabaseConnection()
+        self.reader = DatabaseRecordReader(self.fake_connection)
+
+    def test_row_factory_set(self):
+        self.assertEqual(self.fake_connection.row_factory,
+                         self.reader._row_factory)
+
+    def test_yield_latest_records(self):
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: We might want to flesh these out.

File: tests/unit/customizations/history/test_db.py [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+        reloaded = json.loads(string_value)
+        self.assertEqual(original, reloaded)
+
+    def test_can_serialize_datetime(self):
+        now = datetime.datetime.now()
+        iso_now = now.isoformat()
+        string_value = json.dumps(now, cls=PayloadSerializer)
+        reloaded = json.loads(string_value)
+        self.assertEqual(iso_now, reloaded)
+
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: Should we add a test for the unknown types?

File: tests/unit/customizations/history/test_db.py [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+
+class ThreadedRecordWriter(object):
+    def __init__(self, writer, fake_connection):
+        self._read_q = queue.Queue()
+        self._write_q = queue.Queue()
+        self._thread = threading.Thread(
+            target=self._threaded_record_writer,
+            args=(writer, fake_connection))
+        self._thread.start()
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: I would break this out to its own separate `start()` method. It is a little strange to spin up the actual thread in the initializer.

File: tests/unit/customizations/history/test_db.py [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+        identifier, request_id = written_record[:2]
+        self.assertTrue(self.UUID_PATTERN.match(identifier))
+        self.assertTrue(self.UUID_PATTERN.match(request_id))
+
+    def test_can_write_http_request_record(self):
+        insert_stmt, written_record = self._write_record({
+            'event_type': 'HTTP_REQUEST',
+            'payload': {
+                'method': 'GET',
+                'headers': {},
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: Should this use the `CaseInsensitiveDict` class?

File: tests/unit/customizations/history/test_db.py [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+                'streaming': False,
+                'headers': {},
+                'body': '...',
+                'status_code': 200,
+                'request_id': '1234abcd'
+            }))
+        )
+        identifier, request_id = written_record[:2]
+        self.assertTrue(self.UUID_PATTERN.match(identifier))
+
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: Do we want to assert for uid patttern on request id's?

File: tests/functional/history/test_db.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        ])
+
+        # Since the foo and bar events were written by the writer_a they all
+        # share an id. foo was written at time 3 which makes it the most
+        # recent, so when we call get_latest_records we should get the
+        # foo and bar records only.
+        reader = DatabaseRecordReader(self.connection)
+        records = set([record['event_type'] for record
+                       in reader.yield_latest_records()])
+        self.assertEqual(set(['foo', 'bar']), records)
```
Comment (kyleknap) [inline tests/functional/history/test_db.py]: Should we add tests to make sure the `yield_latest_records` and `yield_records` methods return items by timestamps?

File: tests/functional/history/test_db.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        }
+        records_to_write = 40
+        for _ in range(records_to_write):
+            writer.write_record(known_record_fields)
+
+        cursor = self._get_cursor()
+        cursor.execute("SELECT COUNT(*) FROM records;")
+        num_records = cursor.fetchone()
+        self.assertEqual(num_records[0], records_to_write)
+
```
Comment (kyleknap) [inline tests/functional/history/test_db.py]: I wonder if it makes sense to put some thread property based testing here in the functional tests. The idea is you would use a single db connection behind a lock and then have threads write to the db and at the end you would take the reader to make sure that all of the records were there and there was X distrinct request id's where X would be the number of http lifecycles multiplied by number of threads. Furthermore, you could do some more testing to make sure that when if you group them together by request id's all of the records are as expected.

File: tests/unit/customizations/history/test_db.py (hunk 3 above)
Comment (stealthycoin) [reply tests/unit/customizations/history/test_db.py]: Yep.

File: tests/unit/customizations/history/test_db.py (hunk 2 above)
Comment (stealthycoin) [reply tests/unit/customizations/history/test_db.py]: Missed these somehow.

File: tests/unit/customizations/history/test_db.py (hunk 4 above)
Comment (stealthycoin) [reply tests/unit/customizations/history/test_db.py]: Yea I think also each test for threads may want a different number of threads so I think ill add a `_spin_up_n_threads()` type function that collects them all into a list which is joined on in the `tearDown()`

File: tests/unit/customizations/history/test_db.py (hunk 6 above)
Comment (stealthycoin) [reply tests/unit/customizations/history/test_db.py]: I can't unless there was a prior API_CALL made. I can either change that logic or have an API_CALL be made first for all these tests.

File: tests/unit/customizations/history/test_db.py (hunk 5 above)
Comment (stealthycoin) [reply tests/unit/customizations/history/test_db.py]: yes

File: tests/functional/history/test_db.py (hunk 1 above)
Comment (stealthycoin) [reply tests/functional/history/test_db.py]: Yea this was a little annoying but I didn't see any reason to ever need that outside tests so I left it as is. In the event that is ever used in practice it could easy lead to colliding ids in a database which means you would start picking up records that didn't belong together.

File: tests/functional/history/test_db.py (hunk 8 above)
Comment (stealthycoin) [reply tests/functional/history/test_db.py]: Yea that sounds good to me. I only have the one thread related test so there could definitely be a lot more.

File: awscli/customizations/history/db.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        return request_id
+
+    def _start_http_lifecycle(self):
+        thread_id = threading.current_thread().ident
+        new_identifier = str(uuid.uuid4())
+        self._request_ids[thread_id] = new_identifier
+
+    def _get_identifier(self):
+        if self._identifier is None:
+            self._identifier = str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: Yes my thought is that if we ever want to introduce multiple uids for some reason we can make multiple writers.

File: awscli/customizations/history/db.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        self._connection.commit()
+
+    def _get_request_id(self, event_type):
+        if event_type == self._START_OF_REQUEST_LIFECYCLE_EVENT:
+            self._start_http_lifecycle()
+        if event_type in self._REQUEST_LIFECYCLE_EVENTS:
+            thread_id = threading.current_thread().ident
+            request_id = self._request_ids.get(thread_id, '')
+        else:
+            request_id = ''
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: Like put a None so its a null value in the database rather than an empty string? That seems reasonable.

File: awscli/customizations/history/db.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        db_record = self._create_db_record(record)
+        cursor.execute(self._WRITE_RECORD, db_record)
+        self._connection.commit()
+
+    def _get_request_id(self, event_type):
+        if event_type == self._START_OF_REQUEST_LIFECYCLE_EVENT:
+            self._start_http_lifecycle()
+        if event_type in self._REQUEST_LIFECYCLE_EVENTS:
+            thread_id = threading.current_thread().ident
+            request_id = self._request_ids.get(thread_id, '')
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: Just for testing if I insert an HTTP record before inserting an API_CALL record then there will be no thread id. This is the same reason I do not check the thread_id in some of those tests, only in the ones where an API_CALL is inserted first.

I can either:
* Make it a direct access with [] and remove all tests that just insert a solitary http related record without first inserting an API_CALL record.
* Leave it
* Go back to what I was originally doing where if there is no request_id for an http record insertion it generates a new one then uses that one. Leaving API_CALL with the same behavior where it overwrites an existing request_id.

In any case the real thing won't use either behavior that is being used for tests so I don't really like either 1 or 3.

File: awscli/customizations/history/db.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+    _START_OF_REQUEST_LIFECYCLE_EVENT = 'API_CALL'
+
+    def __init__(self, connection=None):
+        if connection is None:
+            connection = DatabaseConnection()
+        self._connection = connection
+        self._request_ids = {}
+        self._identifier = None
+
+    def write_record(self, record):
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: Good call.

File: awscli/customizations/history/db.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+            return encoded
+        except UnicodeDecodeError:
+            b64encoded = self._process_bytes(obj, 'bytes')
+            encoded = self.encode(b64encoded)
+            return encoded
+
+    def default(self, obj):
+        type_name = type(obj).__name__
+        return getattr(
+            self, '_process_%s' % type_name, self._unknown
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: Yea that makes more sense.

File: awscli/customizations/history/db.py [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+        # it can handle a bytes object becausae it can't tell it apart from a
+        # str object. In PY3 this is handled by the _process_bytes method.
+        # For PY2 we have to override the encode method where the encoding of
+        # a bytes-like string will fail, catch the UnicodeDecodeError and
+        # then call _process_bytes and retry the encoding.
+        try:
+            encoded = super(PayloadSerializer, self).encode(obj)
+            return encoded
+        except UnicodeDecodeError:
+            b64encoded = self._process_bytes(obj, 'bytes')
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: No it only will encode the bytes object as base64.

File: awscli/customizations/history/db.py [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+    def _try_to_enable_wal(self):
+        try:
+            cursor = self.cursor()
+            cursor.execute(self._ENABLE_WAL)
+            self._connection.commit()
+        except sqlite3.Error:
+            # This is just a performance enhancement so it is optional. Not all
+            # systems will have a sqlite compiled with the WAL enabled.
+            pass
```
[the remark this answers is not in the document]
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: I can add debug logs yea.

File: awscli/customizations/history/db.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (kyleknap) [reply awscli/customizations/history/db.py]: Ok that is good then. Not sure if there is a test for it, but it would be good to make sure it gets caught.

File: awscli/customizations/history/db.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kyleknap) [reply awscli/customizations/history/db.py]: Yep that is what I was thinking. I always feel like replacing empty string for None is a little ambiguous.

File: awscli/customizations/history/db.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kyleknap) [reply awscli/customizations/history/db.py]: I think I am fine keeping it. It will also be more robust if there was an issue in the future where for whatever reason we did not previously put a request id. Like we do not want to lose the whole record.

File: tests/functional/history/test_db.py (hunk 1 above)
Comment (kyleknap) [reply tests/functional/history/test_db.py]: Yeah that makes sense. It was just a thought I had when I was looking through the tests.

File: tests/unit/customizations/history/test_db.py (hunk 6 above)
Comment (kyleknap) [reply tests/unit/customizations/history/test_db.py]: Ah I think we should just then at least for one of the `API_CALL` tests there is a pattern matching assertion for request_id if it is not already there.

File: awscli/compat.py [hunk 16]
```diff
@@ -37,6 +37,12 @@
     ZIP_COMPRESSION_MODE = zipfile.ZIP_STORED
 
 
+try:
+    import sqlite3
+except ImportError:
+    sqlite3 = None
```
[the remark this answers is not in the document]
Comment (jamesls) [reply awscli/compat.py]: This will happen before debug logs are configured though so I don't think it'd ever show up anywhere.  Plus you log if sqlite3 isn't available when you try to enable the customization.

File: awscli/customizations/history/db.py [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+    )"""
+    _ENABLE_WAL = """PRAGMA journal_mode=WAL"""
+    _DATABASE_NAME = 'history.db'
+    _DEFAULT_DATABASE_FILENAME = os.path.expanduser(
+        os.path.join('~', '.aws', 'cli', 'history', _DATABASE_NAME))
+
+    def __init__(self, db_filename=None):
+        if db_filename:
+            self._db_filename = db_filename
+        elif os.environ.get('AWS_HISTORY_PATH'):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: I don't believe this was discussed.  I'd prefer to not add this env var for now.

File: awscli/customizations/history/db.py [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+      event_type TEXT,
+      timestamp INTEGER,
+      payload TEXT
+    )"""
+    _ENABLE_WAL = """PRAGMA journal_mode=WAL"""
+    _DATABASE_NAME = 'history.db'
+    _DEFAULT_DATABASE_FILENAME = os.path.expanduser(
+        os.path.join('~', '.aws', 'cli', 'history', _DATABASE_NAME))
+
+    def __init__(self, db_filename=None):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: You can default to `_DEFAULT_DATABASE_FILENAME` if we're not going to support the `AWS_HISTORY_PATH` env var.

File: awscli/customizations/history/db.py [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+
+    def default(self, obj):
+        if isinstance(obj, binary_type):
+            return self._encode_binary(obj)
+        elif isinstance(obj, datetime.datetime):
+            return self._encode_datetime(obj)
+        elif isinstance(obj, MutableMapping):
+            return self._encode_mutable_mapping(obj)
+        else:
+            return type(obj).__name__
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: Remind me again why not the repr?

File: awscli/customizations/history/db.py [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+    @row_factory.setter
+    def row_factory(self, row_factory):
+        self._connection.row_factory = row_factory
+
+
+class PayloadSerializer(json.JSONEncoder):
+    def _encode_mutable_mapping(self, obj):
+        return dict(obj)
+
+    def _encode_binary(self, obj):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: This seems hacky.  What were the specific history records that resulted in needing this method?

File: awscli/customizations/history/db.py [hunk 21]
```diff
@@ -0,0 +1,256 @@
...
+            self._db_filename = db_filename
+        elif os.environ.get('AWS_HISTORY_PATH'):
+            self._db_filename = os.path.join(
+                os.environ.get('AWS_HISTORY_PATH'), self._DATABASE_NAME)
+        else:
+            self._db_filename = self._DEFAULT_DATABASE_FILENAME
+
+        if not os.path.isdir(os.path.dirname(self._db_filename)):
+            os.makedirs(os.path.dirname(self._db_filename))
+        self._connection = sqlite3.connect(
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: From what I can tell we're committing after every sql statement so I'd prefer to simplify things by using using auto commit for the isolation level and not bothering with cursors/commits.  This code is already heavily coupled to sqlite3 anyways.

File: awscli/customizations/history/db.py [hunk 22]
```diff
@@ -0,0 +1,256 @@
...
+            request_id,
+            record['source'],
+            event_type,
+            record['timestamp'],
+            json_serialized_payload
+        )
+        return db_record
+
+
+class DatabaseRecordReader(object):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: Where is the b64decode happening that undoes the b64encoding from the Writer?

File: awscli/customizations/history/db.py [hunk 23]
```diff
@@ -0,0 +1,256 @@
...
+        cursor = self._connection.cursor()
+        db_record = self._create_db_record(record)
+        cursor.execute(self._WRITE_RECORD, db_record)
+        self._connection.commit()
+
+    def _get_request_id(self, event_type):
+        if event_type == self._START_OF_REQUEST_LIFECYCLE_EVENT:
+            self._start_http_lifecycle()
+        if event_type in self._REQUEST_LIFECYCLE_EVENTS:
+            thread_id = threading.current_thread().ident
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: tls is the typical way to handle this.  Is there a reason you're not using that?

File: awscli/customizations/history/db.py [hunk 24]
```diff
@@ -0,0 +1,256 @@
...
+    def write_record(self, record):
+        # This method is not threadsafe by itself, it is only threadsafe when
+        # used inside a handler bound to the HistoryRecorder in botocore which
+        # is protected by a lock.
+        cursor = self._connection.cursor()
+        db_record = self._create_db_record(record)
+        cursor.execute(self._WRITE_RECORD, db_record)
+        self._connection.commit()
+
+    def _get_request_id(self, event_type):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: This seems like the wrong level of abstraction to track this at.  I wouldn't expect a database record writer to know anything about request lifecycle events.

I would expect:

* Handler - bridge between history records and the backend storage
* DatabaseWriter - munges the records so they can be stored in the database (payload serialization, json encoding, etc).  Coordinate with:
* DBConnection - low level details of talking to sqlite3, dealing with transactions, etc.

File: awscli/customizations/history/db.py [hunk 25]
```diff
@@ -0,0 +1,256 @@
...
+    def _row_factory(self, cursor, row):
+        d = {}
+        for idx, col in enumerate(cursor.description):
+            val = row[idx]
+            if col[0] == 'payload':
+                val = json.loads(val)
+            d[col[0]] = val
+        return d
+
+    def yield_latest_records(self):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: Small nitpick, can you change all the `yield_` to `iter_`?

File: awscli/customizations/history/db.py (hunk 23 above)
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: I had it that way initially I changed it due to feedback. I can put it back if you'd prefer. The docs for `ident` say its intended use was for keying a dictionary off of threads so I felt it wasn't too gross this way.

File: awscli/customizations/history/db.py (hunk 24 above)
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: My thought here was this is actually somewhat a property of the database given that it has its own column, and so the record writer that writes this particular schema already has to have knowledge about it separate from the simple write payload in a row logic. I could do this logic in botocore and put it as part of the payload.

File: tests/functional/history/test_db.py [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+        cursor.execute("SELECT * FROM records;")
+        record = dict(cursor.fetchone())
+        for col_name, row_value in known_record_fields.items():
+            # Normally our reader would take care of parsing the JSON from
+            # the payload.
+            if col_name == 'payload':
+                record[col_name] = json.loads(record[col_name])
+                self.assertEqual(record[col_name], row_value)
+
+        self.assertIn('id', record.keys())
```
Comment (kyleknap) [inline tests/functional/history/test_db.py]: Usually you do not have to use `keys()`. `in` for a dict means to look in the keys for it already.

File: awscli/customizations/history/db.py [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+            yield row
+
+    def iter_records(self, record_id):
+        cursor = self._connection.cursor()
+        cursor.execute(self._GET_RECORDS_BY_ID, [record_id])
+        for row in cursor:
+            yield row
+
+
+class RecordBuilder(object):
```
Comment (kyleknap) [inline awscli/customizations/history/db.py]: Did not see really a test class for the `RecordBuilder`. Do you think it makes sense to add some specifically related to making sure everything is formatted correctly, it handles payloads properly, and injects timestamps? It looks like you really have a lot of logic related to request_id's but not so much the other aspects of it.

File: tests/unit/customizations/history/test_db.py [hunk 28]
```diff
@@ -617,6 +617,20 @@ def _get_request_id_for_event_type(self, builder, event_type):
         record = builder.build_record(event_type, {'body': b''}, '')
         return record.get('request_id')
 
+    def test_can_process_http_request_with_no_body(self):
+        builder = RecordBuilder()
+        try:
+            builder.build_record('HTTP_REQUEST', {}, '')
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: So it is not guaranteed that a body will be available? I think in my code I assume at least the `'body'` key will be present but maybe set as `None`

File: awscli/customizations/history/db.py [hunk 29]
```diff
@@ -0,0 +1,246 @@
...
+from awscli.compat import sqlite3
+from awscli.compat import ensure_text_type
+
+
+LOG = logging.getLogger('awscli.history.db')
+
+
+class DatabaseConnection(object):
+    _CREATE_TABLE = """
+    CREATE TABLE IF NOT EXISTS records (
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: Can you indent this?  It's easier to read if these don't align with class variables indentation level

File: awscli/customizations/history/db.py [hunk 30]
```diff
@@ -0,0 +1,246 @@
...
+      id TEXT,
+      request_id TEXT,
+      source TEXT,
+      event_type TEXT,
+      timestamp INTEGER,
+      payload TEXT
+    )"""
+    _ENABLE_WAL = """PRAGMA journal_mode=WAL"""
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: Don't need triple quotes here.

File: awscli/customizations/history/db.py (hunk 18 above)
Comment (jamesls) [reply awscli/customizations/history/db.py]: Yes, let's remove this from this PR and just default `db_filename` to `_DEFAULT_DATABASE_FILENAME` for now.

File: awscli/customizations/history/db.py [hunk 31]
```diff
@@ -0,0 +1,246 @@
...
+    def __init__(self, db_filename=None):
+        if db_filename:
+            self._db_filename = db_filename
+        elif os.environ.get('AWS_HISTORY_PATH'):
+            self._db_filename = os.path.join(
+                os.environ.get('AWS_HISTORY_PATH'), self._DATABASE_NAME)
+        else:
+            self._db_filename = self._DEFAULT_DATABASE_FILENAME
+
+        if self._db_filename != ':memory:' and \
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: If you have to have hard coded string checks like this, it's a pretty good indication that this is happening at the wrong level of abstraction.  I would prefer you move the makedirs stuff out of the db and let the db just deal with talking to sqlite3.  The thing that accepts a db_filename needs to ensure that this directory exists.  Then you can remove the `:memory:` check.

File: awscli/customizations/history/db.py [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+           not os.path.isdir(os.path.dirname(self._db_filename)):
+            os.makedirs(os.path.dirname(self._db_filename))
+        self._connection = sqlite3.connect(
+            self._db_filename, check_same_thread=False, isolation_level=None)
+        self._ensure_database_setup()
+
+    def execute(self, query, *parameters):
+        self._connection.execute(query, *parameters)
+
+    def cursor(self):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: We discussed removing the cursor didn't we?  I'd suggest simplifying this interface and not exposing cursors at all.  We shouldn't need it.

File: awscli/customizations/history/db.py [hunk 33]
```diff
@@ -0,0 +1,246 @@
...
+    def _create_record_table(self):
+        self.execute(self._CREATE_TABLE)
+
+    def _try_to_enable_wal(self):
+        try:
+            self.execute(self._ENABLE_WAL)
+        except sqlite3.Error:
+            # This is just a performance enhancement so it is optional. Not all
+            # systems will have a sqlite compiled with the WAL enabled.
+            pass
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: Can you log this at debug?

File: awscli/customizations/history/db.py [hunk 34]
```diff
@@ -0,0 +1,246 @@
...
+
+
+class PayloadSerializer(json.JSONEncoder):
+    def _encode_mutable_mapping(self, obj):
+        return dict(obj)
+
+    def _encode_datetime(self, obj):
+        return obj.isoformat()
+
+    def _unknown(self, obj, type_name):
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: Where is this used?

File: awscli/customizations/history/db.py [hunk 35]
```diff
@@ -0,0 +1,246 @@
...
+            return self._encode_datetime(obj)
+        elif isinstance(obj, MutableMapping):
+            return self._encode_mutable_mapping(obj)
+        else:
+            return repr(obj)
+
+
+class DatabaseRecordWriter(object):
+    _WRITE_RECORD = """
+    INSERT INTO records(id, request_id, source, event_type, timestamp, payload)
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: indent please.

File: awscli/customizations/history/db.py [hunk 36]
```diff
@@ -0,0 +1,246 @@
...
+            return request_id
+        return None
+
+    def _format_payload(self, event_type, payload):
+        # If the payload has a body key it can be in bytes, so it needs to be
+        # converted to utf-8 if possible so it can be serialized as JSON later.
+        if event_type in self._BYTES_BODY_PAYLOADS:
+            body = payload.get('body')
+            if body is not None:
+                payload['body'] = ensure_text_type(body)
```
Comment (jamesls) [inline awscli/customizations/history/db.py]: I don't think you answered my original question about this (I might have missed it though).  What specific instances are you running into for this?  Additionally, if a payload can be bytes, then it's possible that you can't utf-8 encode it.  It looks like you just removed the previous version's base64 encoding.  Are you certain we don't need it?

Comment (stealthycoin) [reply awscli/customizations/history/db.py]: The body key on http_request and http_response payloads is as far as I can tell always bytes if it is not None.

File: awscli/customizations/history/db.py (hunk 32 above)
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: For writing yes it uses the connection and autocommit mode. I do not see a way to read directly from the connection though, the __iter__ and all the fetch* commands are on the cursor only as far as I can tell.

Comment (jamesls) [reply awscli/customizations/history/db.py]: It's exactly the same as inserts: https://docs.python.org/2/library/sqlite3.html#using-shortcut-methods

File: awscli/customizations/history/db.py [hunk 37]
```diff
@@ -25,38 +25,25 @@
...
+        CREATE TABLE IF NOT EXISTS records (
+          id TEXT,
+          request_id TEXT,
+          source TEXT,
+          event_type TEXT,
+          timestamp INTEGER,
+          payload TEXT
+        )"""
+    _ENABLE_WAL = 'PRAGMA journal_mode=WAL'
+    _DEFAULT_DATABASE_NAME = 'history.db'
```
Comment (kyleknap) [inline awscli/customizations/history/db.py]: Can probably remove this class variable now.

File: tests/unit/customizations/history/test_db.py [hunk 38]
```diff
@@ -607,81 +599,89 @@ def test_can_serialize_unknown_type(self):
...
+
+    def test_does_create_record_with_correct_fields(self):
+        record = self.builder.build_record('type', 'payload', 'source')
+        self.assertEqual(record['event_type'], 'type')
+        self.assertEqual(record['payload'], 'payload')
+        self.assertEqual(record['source'], 'source')
+        self.assertTrue('command_id' in record)
+        self.assertTrue('timestamp' in record)
+
+    def test_can_process_http_request_with_noone_body(self):
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: typo: noone --> none

File: tests/unit/customizations/history/test_db.py [hunk 39]
```diff
@@ -607,81 +599,89 @@ def test_can_serialize_unknown_type(self):
...
         try:
-            builder.build_record('HTTP_REQUEST', {}, '')
+            self.builder.build_record('HTTP_REQUEST', {'body': None}, '')
         except ValueError:
             self.fail("Should not raise value error")
 
-    def test_can_process_http_response_with_no_body(self):
-        builder = RecordBuilder()
+    def test_can_process_http_response_with_nono_body(self):
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: Remove this test. It is the copy of the last one now.

File: awscli/customizations/history/db.py [hunk 40]
```diff
@@ -83,11 +79,53 @@ def _encode_mutable_mapping(self, obj):
     def _encode_datetime(self, obj):
         return obj.isoformat()
 
+    def _try_encode(self, obj):
```
Comment (kyleknap) [inline awscli/customizations/history/db.py]: Shouldn't this be `_try_decode()`? Maybe call it `_try_decode_bytes()`?

File: awscli/customizations/history/db.py [hunk 41]
```diff
@@ -83,11 +79,53 @@ def _encode_mutable_mapping(self, obj):
...
+        try:
+            obj = obj.decode('utf-8')
+        except UnicodeDecodeError:
+            obj = '<Byte sequence>'
+        return obj
+
+    def _remove_non_unicode_stings(self, obj):
+        if isinstance(obj, str):
+            obj = self._try_encode(obj)
+        elif isinstance(obj, dict):
```
Comment (kyleknap) [inline awscli/customizations/history/db.py]: Hmm. Would it be possible to do this in place or would does that risk mutating the payload in the upstream code logic?

File: awscli/customizations/history/db.py [hunk 42]
```diff
@@ -83,11 +79,53 @@ def _encode_mutable_mapping(self, obj):
...
 
+    def _try_encode(self, obj):
+        try:
+            obj = obj.decode('utf-8')
+        except UnicodeDecodeError:
+            obj = '<Byte sequence>'
+        return obj
+
+    def _remove_non_unicode_stings(self, obj):
+        if isinstance(obj, str):
```
Comment (kyleknap) [inline awscli/customizations/history/db.py]: What's your thoughts on using `binary_type` here instead?

File: tests/unit/customizations/history/test_db.py [hunk 43]
```diff
@@ -598,6 +538,64 @@ def test_can_serialize_unknown_type(self):
...
+    def test_can_serialize_non_utf_8_bytes_type_in_list(self):
+        original = ['foo', b'\xfe\xed']
+        encoded = ['foo', '<Byte sequence>']
+        string_value = json.dumps(original, cls=PayloadSerializer)
+        reloaded = json.loads(string_value)
+        self.assertEqual(encoded, reloaded)
+
+    def test_can_serialize_non_utf_8_bytes_type_in_tuple(self):
+        original = ('foo', b'\xfe\xed')
+        encoded = ['foo', '<Byte sequence>']
```
Comment (kyleknap) [inline tests/unit/customizations/history/test_db.py]: Should this remain a tuple?

File: tests/functional/history/test_db.py [hunk 44]
```diff
@@ -301,3 +299,168 @@ def test_can_read_most_recent_records(self):
...
+        '^[0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12}$',
+        re.I
+    )
+
+    def setUp(self):
+        self.db = DatabaseConnection(':memory:')
+        self.writer = DatabaseRecordWriter(connection=self.db)
+        self.handler = DatabaseHistoryHandler(writer=self.writer)
+
+    def _assert_last_record_structure(self, event_type, payload, source,
```
Comment (kyleknap) [inline tests/functional/history/test_db.py]: It's kind of hard to follow this. I honestly would prefer there be the following:

* `_get_last_record()` to get the last record and the payload already loaded.
* `_assert_record_is_correct_structure()` which does the assertion on the tuple
* Then have more helper assertion methods like `_assert_expected_event_type()`, `_assert_expected_source_type()`, and `_assert_expected_payload()`, `_assert_has_request_id()`

The problem now for me is the assertion is too vague as it is now and makes it hard to follow the assertions.

File: tests/unit/customizations/history/test_db.py (hunk 43 above)
Comment (stealthycoin) [reply tests/unit/customizations/history/test_db.py]: No its being serialzied to JSON which turns it into a list.

File: awscli/customizations/history/db.py (hunk 42 above)
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: I match exactly what the JSON encoder does so I would be hesitant to deviate.

File: awscli/customizations/history/db.py (hunk 41 above)
Comment (stealthycoin) [reply awscli/customizations/history/db.py]: That would probably mutate it. I added a few tests to check for mutation.
